server/src/main/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenStore.java

```diff
@@ -83,24 +84,36 @@ public class UaaTokenStore implements AuthorizationCodeServices {
     private final RandomValueStringGenerator generator = new RandomValueStringGenerator(32);
     private final RowMapper rowMapper = new TokenCodeRowMapper();
 
-    private AtomicReference<Instant> lastClean = new AtomicReference<>(Instant.EPOCH);
+    private final AtomicReference<Instant> lastClean = new AtomicReference<>(Instant.EPOCH);
+
+    private final Clock clock;
 
     public UaaTokenStore(DataSource dataSource) {
         this(dataSource, DEFAULT_EXPIRATION_TIME);
     }
 
     public UaaTokenStore(DataSource dataSource, Duration expirationTime) {
+        this(dataSource, expirationTime, Clock.systemDefaultZone());
+    }
+
+    private UaaTokenStore(DataSource dataSource, Duration expirationTime, Clock clock) {
         this.dataSource = dataSource;
         this.expirationTime = expirationTime;
+        this.clock = clock;
+    }
+
+    static UaaTokenStore constructorForTesting(DataSource dataSource, Duration expirationTime, Clock clock) {
+        return new UaaTokenStore(dataSource, expirationTime, clock);
     }
 
     @Override
     public String createAuthorizationCode(OAuth2Authentication authentication) {
-        final int max_tries = 3;
+        final int maxAttempts = 3;
         performExpirationClean();
         JdbcTemplate template = new JdbcTemplate(dataSource);
-        int tries = 0;
-        while ((tries++)<=max_tries) {
+        int attempt = 0;
+        while (true) {
```

Comment:
The check on maxTries has moved inside a catch block. Is there a potential infinite loop here?

Reply:
It hasn't moved. It disappeared. As I explain in the commit message, this tests is always `true`, as explained by IntelliJ.

Reply:
Although the change is not going to cause an infinite loop because of the `throw` in the `catch` block, this change does modify the code behavior.
The old code will gracefully exit the while loop when `tries` becomes 3.  The new code will throw `DataIntegrityViolationException`.   Notice that the post-increment operator (`++`) was used in the condition in the old code.

Reply:
>Part of the method is not well-written

What was your first clue?

Reply:
>The throw and rethrow of DataIntegrityViolationException is unnecessary.


I agree, this is pretty crazy.

Are you saying that you want this changed as part of this PR?

Reply:
> > Part of the method is not well-written
> 
> What was your first clue?

I could only see what the old code was doing after expanding the hidden code block in this PR as that code was not part of this PR change.

Reply:
Bruce is right, this was already in the catch block.

But I do not see how `(tries++)<=max_tries` is always true, and the refactor that my IntelliJ suggests that the logic move inside the loop, not that it be removed entirely. I'm still concerned about a potential infinite loop, and I'm curious why there isn't a test that catches it.

Reply:
I'm gonna makes more PRs on this code anyways.

Reply:
I reverted that change. I'll make another short PR for it.